autoscale-groups-core/src/main/java/fish/payara/extensions/autoscale/groups/core/admin/ScaleCommandHelper.java

```diff
@@ -0,0 +1,327 @@
+
+            String msg = commandName + " " + serverName;
+            LOGGER.info(msg);
+
+            // Wrap the command invocation in a runnable and hand it off
+            // to the thread pool
+            CommandRunnable cmdRunnable = new CommandRunnable(invocation, instanceReport, responseQueue);
+            cmdRunnable.setName(serverName);
+            threadPool.execute(cmdRunnable);
+        }
+
+        if (LOGGER.isLoggable(FINE)) {
+            LOGGER.fine(String.format("%s commands queued, waiting for responses", commandName));
+        }
+
+        // Make sure we don't wait longer than the admin read timeout. Set our limit to be 3 seconds less.
+        long adminTimeout = RemoteRestAdminCommand.getReadTimeout() - 3000;
+        if (adminTimeout <= 0) {
+            // This should never be the case
+            adminTimeout = 57 * 1000;
+        }
+
+        if (LOGGER.isLoggable(FINE)) {
+            LOGGER.fine(String.format("Initial cluster command timeout: %d ms", adminTimeout));
+        }
+
+        // Now go get results from the response queue.
+        for (int n = 0; n < nInstances; n++) {
+            long timeLeft = adminTimeout - (System.currentTimeMillis() - startTime);
+            if (timeLeft < 0) {
+                timeLeft = 0;
+            }
+            CommandRunnable cmdRunnable = null;
+            try {
+                cmdRunnable = responseQueue.poll(timeLeft, MILLISECONDS);
+            } catch (InterruptedException e) {
+                // This thread has been interrupted. Abort
+                threadPool.shutdownNow();
+                String msg = Strings.get("cluster.command.interrupted", targetNames, n, nInstances, commandName);
+                LOGGER.warning(msg);
+                failureOccurred = true;
+                // Re-establish interrupted state on thread
+                Thread.currentThread().interrupt();
+                break;
+            }
+
+            if (cmdRunnable == null) {
+                // We've timed out.
+                break;
+            }
+            String cname = cmdRunnable.getName();
+            waitingForServerNames.remove(cname);
+            ActionReport instanceReport = cmdRunnable.getActionReport();
+            if (LOGGER.isLoggable(FINE)) {
+                LOGGER.fine(String.format("Instance %d of %d (%s) has responded with %s", n + 1, nInstances, cname,
+                        instanceReport.getActionExitCode()));
+            }
+
+            if (instanceReport.getActionExitCode() != SUCCESS) {
+                // Bummer, the command had an error. Log and save output
+                failureOccurred = true;
+                failedServerNames.append(cname).append(" ");
+                reportResult.failedServerNames.add(cname);
+                String msg = cname + ": " + instanceReport.getMessage();
+                LOGGER.severe(msg);
+                msg = Strings.get("cluster.command.instancesFailed", commandName, cname);
```

Comment:
This may reference the wrong message string, see my previous comment

Reply:
`The command {0} failed for: {1}`